### DISCUSSION TITLE
When running service using shawl with pyinstaller, extracted files are owned by Administrator not designated user

### DISCUSSION THREAD

**Discussion by nkmhor** [pyinstaller/pyinstaller#6920] [Help]
I am using shawl to run a python program as a Windows service.  The python code is wrapped up into an `.exe` using `pyinstaller`.  To create the service I execute something like the following from an elevated command (Administrator) prompt:

`sc create ServiceName start=auto obj=myuser password="password" binPath = "C:\Users\me\shawl.exe run --cwd d:\wd -- c:\Users\me\python_prog.exe"`

The problem is that the files unpacked from `python_prog.exe` are owned by Administrator and not the myuser specified with `sc`.  One of the files is an Excel spreadsheet and when the python code, running as myuser tries to access the spreadsheet it can't because the spreadsheet is owned by Administrator.   Is this a result of shawl or of pyinstaller? The files unpacked by pyinstaller are placed in `C:\Users\myuser\AppData\Local\Temp`.  So if the files are unpacked in myuser by `pyth_prog.exe` running inside `shawl.exe` running as myuser, why aren't they owned by myuser?  If I supply a `--runtime-temp` argument to pyinstaller, then the files are unpacked where I specify but they are still owned by Administrator.
```
OS: Windows Server 2019 Datacenter
Python: 3.9.5
pyinstaller: 4.7
shawl: 1.1.0
```

**Comment by rokm** [comment]
What is the PyInstaller command you are using to build your .exe? Are you using `--uac-admin` switch?

**Reply by nkmhor** [reply]
I am not using `--uac-admin`:
`pyinstaller --onefile --paths="mypath" --hidden-import=win32timezone --add-data="mydata" --name my.exe my.py`

**Reply by rokm** [reply]
And if you run `my.exe` as a user manually, who is the owner of the unpacked temporary directory?

**Reply by nkmhor** [reply]
The principal that ran my.exe, i.e. the files are owned by myuser if run as myuser but if I login as myuser and open an elevated command prompt and then run my.exe then the files are owned by Administrator.

**Reply by rokm** [reply]
So my guess would be that `my.exe` is being ran as Administrator.

**Reply by nkmhor** [reply]
I have a difficulty seeing how that happens given that sc is told to run the service as user myuser and the files are unpacked in \Users\myuser\AppData\Local\tmp.  Also in task manager, my.exe shows a username of myuser.

**Reply by rokm** [reply]
If I run a sample onefile program from elevated command prompt, the files are extracted into `C:\Users\Rok\AppData\Local\Temp\_MEIXXXXX`, the Task manager shows that they are ran under my user (Rok), but the files are owned by Administrators.

**Reply by nkmhor** [reply]
Right, but I am not running the onefile interactively from an elevated command prompt.  I am running sc which should start a service as user myuser and then run shawl.exe (as myuser) which, in turn, should run onefile as myuser.

**Reply by bwoodsend** [reply]
Reading [shawl's readme](https://github.com/mtkennerly/shawl#shawl):

> Bear in mind that the default account for new services is the Local System account,

**Reply by nkmhor** [reply]
That is why I am specifying a user (`obj=`) with the sc command above.  I do not want this service to run as Local System.

**Reply by rokm** [reply]
> Right, but I am not running the onefile interactively from an elevated command prompt. I am running sc which should start a service as user myuser and then run shawl.exe (as myuser) which, in turn, should run onefile as myuser.

My point is, the symptoms you describe are the ones of the executable being ran in elevated mode, even if you have no reason to believe that this is the case.

Even on Windows, I doubt it is possible for an unprivileged process to create Administrator-owned files, due to security implications.

As a simple test, I wrote a simple C program that opens a file and periodically writes to it. Then set up the service via `shawl.exe` like you described, setting `obj` to my local user. If I run this service, the Task Manager shows that the program is running under my username, but the created file is again owned by `HOSTNAME\Administrators`.

So this has nothing to do with PyInstaller at all...

**Reply by rokm** [reply]
Also, in Task Manager, you can go to Details tab, right click on the list-view's header, and click `Select columns`. There, choose `Elevated` to see what processes are elevated - in the above example, both `shawl.exe`and the target executable are marked as elevated, as expected by file ownership.

**Reply by nkmhor** [reply]
Rok,  Thanks for going "above and beyond" in investigating this.  There is some arcana of Windows Services that I will need to read up on to understand why services run as a named user create files as Administrator.
